% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ldsc-rlm1.R
\name{GCvr}
\alias{GCvr}
\title{A robust and efficient LD score regression approach to estimating the genetic correlation using GWAS summary data}
\usage{
GCvr(Zs, r2, N1, N2, Nc = 0, W = NULL)
}
\arguments{
\item{Zs}{Mx2 matrix of summary Z-statistics for M variants from two GWAS}

\item{r2}{average reference LD scores for M variants}

\item{N1}{sample size for the 1st GWAS}

\item{N2}{sample size for the 2nd GWAS}

\item{Nc}{overlapped sample size between the two GWAS}

\item{W}{variant weight}
}
\value{
\describe{
  \item{gv}{ genetic covariance }
  \item{gc}{ genetic correlation }
  \item{r0}{ estimated intercept, quantifying the marginal trait correlation }
  \item{h2s}{ SNP heritabilities }
}
}
\description{
We follow the Bulik-Sullivan et al. (2015) approach but use a robust LS estimation. 
The input r2 is the averaged reference LD scores (scaled by the total number of SNPs used to compute the LD scores).
The weight W is typically based on the LD scores computed using HapMap3 common SNPs to correct over-counting.
}
\references{
Bulik-Sullivan,B.K. et al. (2015) An atlas of genetic correlations across human diseases and traits. Nat. Genet. 47, 1236â€“1241.

Guo,B. and Wu,B. (2018) Principal component based adaptive association test of multiple traits using GWAS summary statistics. bioRxiv 269597; doi: 10.1101/269597
}
